modernsyntax/classical_looping.cpp

```diff
@@ -28,7 +28,15 @@ void indexLoop(std::vector<Tool>& tools){
 }
 
 void iteratorLoop(std::vector<Tool>& tools){
-  for(std::vector<Tool>::const_iterator tool = tools.begin(), end = tools.end(); tool != end; ++tool){
+  for(auto tool = tools.begin(), end = tools.end(); tool != end; ++tool){
+    tool->print();
+  }
+}
+
+void rangeLoop(std::vector<Tool>& tools, const int beginning, const int ending){
+  auto tool = tools.begin();
+  std::advance(tool,beginning);
+  for(auto end = tools.end(); std::distance(tool, end) > std::distance(tools.begin(), end) - ending; ++tool){
```

Comment:
Here `std::distance` is being calculated during every loop step. Is there a more performant way?

Reply:
The second distance could definitely be stored. The first one might behave differently if the loop content alters the size of `tools` or shifts `tool`. For our simple print this is of course no concern so a full pre-calculation would be more efficient.

Reply:
Yap. That is completely correct.